### ISSUE TITLE
Add option to always include the country code in the text field

### ISSUE THREAD

**Issue by TylerRick** [catamphetamine/react-phone-number-input#273]
*Parsing* the country code *from* the text field (some of the time) without also always *displaying* it in the text field can lead to a frustrating/confusing UX.

Making the country code always visible would provide an extra feedback to the user as to whether the number they entered is what they intended to enter.

Example of a possible confusing user interactions (I used this [codesandbox](https://codesandbox.io/s/react-phone-number-input-bad-ux-demo-3eqyz)):

*   Let's say I don't realize that a country code is required, so I just start typing a phone number as it would normally be entered in my country (US), like `206-555-1234`. But to my surprise, Egypt is selected as the country (it was parsed _from_ the text field) and the value is formatted as `+20 65 551234`. If I as the user am paying close attention, this should clue me in to the fact that it did _not_ interpret my number as a US number. But since it is formatted all strangely now, it may not be clear exactly _how_ to fix the problem. I might try one of 2 things:
* `1.` Try to fix it via the text field (keyboard)
    *   First, maybe I clear the input by backspacing. Egypt is still selected as the country, but since no country code is visible to the user, I might assume I've cleared the input completely (resetting it back to its initial state).
    * Now let's say I either the same number again, `206-555-1234`, or I try something different and add a `1`, `1-206-555-1234` (many people don't know you need a `+` before the country code) . Either way, I get a wrong result (either `"+202065551234"` or `"+2012065551234"` as the new internal state, respectively), but worst of all, may not realize it because the actual result is not visible to the user! What I _see_ in the field is `2065551234`, which at least looks better (more normal) than `+20 65 551234`, so I might conclude that it interpreted the number correctly this time (wrong!) and proceed with filling in the rest of the form. Unless I've switched to my mouse and actually _clicked_ on the flag, I still might not realize that the flag is somehow an integral _part_ of the phone number field. And even if I _do_ click the drop-down, since it doesn't show the country codes next to the countries (requested in #259 and :+1: by me), nor in the text field, it's not clear that they will actually be included in the value that gets submitted by the form, and I may still believe it's going to submit it correctly.
* Or `2.` Try to fix it by choosing a different country from the select (mouse)
    - Perhaps just as likely, I, the hapless user, notice that the flag is incorrect and click on it to discover the drop-down. Sure enough, I see that Egypt is selected. "That's not right! Let me fix that..."
    - So I choose United States from the list and ... wat? Now it says `+1 655 512 34`! What happened to the `206` area code that I entered? (And of course, to add insult to injury, if I start  typing `20` after the `+1`, I'll run into the issue where only the `2` ends up where I want it and the `0` ends up at the end.)

Related issues: #52, #201

## Either always read/show country code from/in text field or never do

Currently it is inconsistent: the first time they enter a number, it interpreting it as a full, _international_ number (taking the first digits as the country code), but after a country has been selected once, it interpreted _all_ of the digits as being a _national_ number. 

I can understand how this was probably the intended behavior, and how it _can_ be a pleasant UI some of the time. Since the default option is "International", it makes sense that it would parse inputs as international numbers then but then switch to parsing as national numbers as soon as they select a country. Especially when you can set a default country, this might actually be a better UX because most people would probably _prefer_ to see their phone number in their local/national format, right? 

But there are 3 problems with that:
- People may not click on the country picker at all and not _realize_ that the default selected option is "International".
- As far as us knowing it's a national number because they selected a country... That would certainly be true *if* the user were required to actually select it _from the list_. But if all they did was enter some digits, and we _assumed_ some of them were the country code, then no, they haven't told us what their country is definitely; we just guessed it.
- To be consistent, after parsing the country code out of the digits and setting the country, in order to properly switch from an international to a national phone number input, it seems like it should _remove_ those country digits from the text field so that the remaining digits _are_ only the digits that are for a national number in the selected country. (This goes back to my main point about how it should either always show or never show the country code in the text field.)

## Related feature request: `countryCodeEditable` prop

To make it even more fool-proof, maybe there should be the option to not only have the country code always be visible, but also not be deletable, like you can do with [`react-phone-input-2`](https://github.com/bl00mber/react-phone-input-2) ([demo](https://schoolbus.netlify.com/?path=/story/input-tel-input--default) ([source](https://github.com/lessondesk/schoolbus/blob/master/src/components/tel-input/index.js))). Seems like a pretty good option if you want to force people to use the country select to set the country code.

But if we allow country code to be editable in the field, then it seems like editing it there should cause it to select the corresponding country/flag. It looks like you _can_ edit the country code _some_ of the time with `react-phone-number-input` — as long as the `+` prefix is still there — but if you accidentally or purposefully remove the `+` then you can't edit the country code anymore, which could be confusing. I really like how [`react-phone-input-2`](https://github.com/bl00mber/react-phone-input-2) won't even let you remove the `+` — even with `countryCodeEditable={true}`.

But both of these behaviors depend on having the country code always visible/included.

--- 
Another idea would be to add an option to not include an "International" option in the select. Then it would only have to parse national numbers.

---
In conclusion, this looks like an awesome library! The hybrid international/national number field idea seems to have some advantages, but also seems like it could possibly be confusing.

I think I may go with [react-phone-input-2](https://github.com/bl00mber/react-phone-input-2) for now, but there are some things about this library that I really like, too (such as the ability to _not_ include the flags in the drop-down so that it loads faster and is more responsive), so it's still in the running...

**Comment by catamphetamine** [comment]
Wow, that's a lot of text.

**Comment by catamphetamine** [comment]
> Let's say I don't realize that a country code is required, so I just start typing a phone number as it would normally be entered in my country (US), like 206-555-1234. But to my surprise, Egypt is selected as the country (it was read from the text field) and the value is formatted as +20 65 551234.

Yes, because "International" is selected.
If you want better UX then employ IP country auto-detection and pass the `country` property.

> If I as the user am paying close attention, this should clue me in to the fact that it did not interpret my number as a US number. But since it is formatted all strangely now, it may not be clear exactly how to fix the problem.

Yes, and the only solution seems to be country autodetection.

> First, maybe I clear the input by backspacing. Egypt is still selected as the country, but since no country code is visible to the user, I might assume I've cleared the input completely (resetting it back to its initial state).

Yeah, I guess.
In the case of clearing it completely selecting "International" would be the best behavior but that would require keeping the history and analysing it on Backspace and I don't want to do that due to the complexity.
You can fork and release your own version though.

> Now let's say I either the same number again, 206-555-1234, or I try something different and add a 1, 1-206-555-1234 (many people don't know you need a + before the country code) . Either way, I get a wrong result

I didn't proceed reading that case you've described.
Either the user sees the flag and realises that it's not US (every US citizen knows the US flag because it's hanging on every door) so they click the flag and see the country list and then they get the idea.
Or the user doesn't get it and not much could be done at that point other than "that's how life is".

> Or 2. Try to fix it by choosing a different country from the select (mouse)
Perhaps just as likely, I, the hapless user, notice that the flag is incorrect and click on it to discover the drop-down. Sure enough, I see that Egypt is selected. "That's not right! Let me fix that..."
So I choose United States from the list and ... wat? Now it says +1 655 512 34! What happened to the 206 area code that I entered? (And of course, to add insult to injury, if I start typing 20 after the +1, I'll run into the issue where only the 2 ends up where I want it and the 0 ends up at the end.)

Yeah, that's a valid statement.
The currently impelmented logic is that it extracts the "national (significant) number" and applies it to the US.
What you're suggesting seems to be extracting all digits and applying them for the US discarding the `+` sign.
I'd say it's the more correct behavior and should be corrected in the library.
Not in the near future though because I currently have my hands full.
Maybe in Autumn.
Even if someone submits a PR it won't be merged until I'm done with my current business.

> Currently it is inconsistent: the first time they enter a number, it interpreting it as a full, international number (taking the first digits as the country code), but after a country has been selected once, it interpreted all of the digits as being a national number.

Yes, and that's because until a country is selected there's no way to interpret the digits being input being anything other than "international".
When a country has been selected (or detected) though it switches into the "intuitive way the stuff works" mode.

> People may not click on the country picker at all and not realize that the default selected option is "International".

I guess most of them don't realise that it's clickable even if it has the arrow.

> As far as us knowing it's a national number because they selected a country... That would certainly be true if the user were required to actually select it from the list. But if all they did was enter some digits, and we assumed some of them were the country code, then no, they haven't told us what their country is definitely; we just guessed it.

Yes, well, the library could track whether the dropdown has been expanded or the `country` property been passed in which case it would be in the "legitimately selected" country mode.
Otherwise it would be in some "auto-selected" country mode which would result in digits being input still being forced to be interpreted as "international". And switching countries in this "international" mode would not strip the "country calling code" as you mentioned above. That would be the only difference between the modes I guess.

> To be consistent, after parsing the country code out of the digits and setting the country, in order to properly switch from an international to a national phone number input, it seems like it should remove those country digits from the text field so that the remaining digits are only the digits that are for a national number in the selected country. (This goes back to my main point about how it should either always show or never show the country code in the text field.)

Hmm, maybe.
That could be a special property, something like `switchToNationalNumberModeAfterDetectingCountry` or something like that.

> This goes back to my main point about how it should either always show or never show the country code in the text field.

It can't be "always international" because there're people not familiar with it (and seems that there's many of them).
And it can't be "always local" because it starts with "International" when the input is empty.

> I really like how react-phone-input-2 won't even let you remove the + — even with countryCodeEditable={true}.

Well, same thing: most people would be confused about what should they input after the `+${code}` stuff so it can't behave like that.
Maybe there could be a special property for that, something like `forceInternational`.
But that would have to be a new feature and is not guaranteed.

> Another idea would be to add an option to not include an "International" option in the select. Then it would only have to parse national numbers.

There is such an option and it's `international={false}`.
See the props docs.

> In conclusion, this looks like an awesome library! The hybrid international/national number field idea seems to have some advantages, but also seems like it could possibly be confusing.

Yeah, it seems relatively good-enough.

> I think I may go with react-phone-input-2 for now, but there are some things about this library that I really like, too (such as the ability to not include the flags in the drop-down so that it loads faster and is more responsive), so it's still in the running...

As I've already said, I'm currently busy in my life so if you're fine with the "as-is" state of things then use this library (which has a couple of non-resolved bugs though, didn't look much into them).

**Comment by TylerRick** [comment]
Thanks for taking the time to reply! I totally understand about being too busy. No rush...

Excellent points all around.

> ... so they click the flag and see the country list and then they get the idea.

Sure... hopefully. Or they press Shift+Tab, Space and stay on their keyboard... :) (I'm so glad it's possible to tab to the country select. For some reason this isn't possible with `react-phone-input-2`, which seems like an accessibility fail.)

> Yes, well, the library could track whether the dropdown has been expanded or the `country` property been passed in which case it would be in the "legitimately selected" country mode.
> Otherwise it would be in some "auto-selected" country mode which would result in digits being input still being forced to be interpreted as "international". And switching countries in this "international" mode would not strip the "country calling code" as you mentioned above. That would be the only difference between the modes I guess.

That sounds like it could work.

> That could be a special property, something like `switchToNationalNumberModeAfterDetectingCountry` or something like that.

Works for me! ... maybe someday... 

Or `stripCountryCodeAfterDetectingCountry` ... if we didn't want users to have to understand different "modes". Either way works for me. But that magic half-way-in-between mode that it currently puts you in currently when it parses a country from input seems a bit too ambiguous to me. So this flag would be a way for developers to opt out of this magic if they wished.

> Yes, and that's because until a country is selected there's no way to interpret the digits being input being anything other than "international".

> And it can't be "always local" because it starts with "International" when the input is empty.

True enough. And thanks for the tip about `international={false}`. Somehow I'd missed that. Wow, you already thought of (almost) everything!

I think the problems I described above were only possible when a user started from "International" mode. 

So, maybe a combination of setting a reasonable default country (such as auto detected from IP address, as you suggested) and `international={false}` will be enough to make this workable for me. Thanks!

> Well, same thing: most people would be confused about what should they input after the `+${code}` stuff so it can't behave like that.
> Maybe there could be a special property for that, something like `forceInternational`.
> But that would have to be a new feature and is not guaranteed.

Works for me. Though I'd suggest calling it `alwaysShowCountryCode`. `forceInternational` might sound like it's forcing the "International" option to be selected.

Would people really be confused? You probably would know better than me. But, it seems like a proper label, helper text, and/or placeholder text would make it clear that an international number was expected. This is the default placeholder text for [`react-phone-input-2`](https://github.com/bl00mber/react-phone-input-2), for example, which gives the user an idea of what's expected before they start typing:

![image](https://user-images.githubusercontent.com/14151/63069516-0059c600-becc-11e9-962f-7c8370663800.png)

I just think there are some cases when the country could be equally likely any of one of the nearby countries — and it wouldn't be good/fair to default to any of them (thus arbitrarily biasing the input towards any one particular country — and increasing the chance that they would leave it set to that country, and not forcing them to pick one, which would actually be better in this case).

When the user is entering their _own_ phone number, I think it's okay to set a default based on the user's IP address. But when they're entering _other_ people's phone numbers, then the country that the user is currently in isn't _as_ safe of a guess. (Especially in Europe, for example, where there are lots of small neighboring countries with people flowing freely between them.)

**Comment by TylerRick** [comment]
> Maybe there could be a special property for that, something like `forceInternational`.  
> But that would have to be a new feature and is not guaranteed.

Should I create a new issue for that or can this issue be about adding that new option?

**Comment by catamphetamine** [comment]
@TylerRick 

I opened this issue today and don't even understand what it's about and what did I write there in the comments above.
I guess this should be restarted.
Too much text, I see it but it's not clear what this all conversation means.
I don't even know what my proposed features were.
I guess this whole ticket should be discarded now.
I'll close it.
If you manage to extract some useful bits from it you may create separate issues for each useful bit.
But don't describe it on several pages like you always do: while informative that's too overwhelming and you see where it leads eventually (nowhere).

Originally I was writing a lengthy comment but then I realized that it'll just push it deeper:

> Let's say I don't realize that a country code is required, so I just start typing a phone number as it would normally be entered in my country (US), like 206-555-1234. But to my surprise, Egypt is selected as the country (it was parsed from the text field) and the value is formatted as +20 65 551234. If I as the user am paying close attention, this should clue me in to the fact that it did not interpret my number as a US number. But since it is formatted all strangely now, it may not be clear exactly how to fix the problem. I might try one of 2 things:

So, I guess the issue is that the input doesn't communicate which format does it require the phone number to be input in.
If "International" is selected and the input is empty then it could show a different placeholder, something like `"Enter your phone number in international format"` except that it's a huge placeholder.
It can't be an example number I guess because every country has a different format of a international phone number so that would confuse people too.
Not to mention that example placeholders themselves confuse people because they think some value is already input in such field (placeholders look liking like actual values).

> so I might conclude that it interpreted the number correctly this time (wrong!) and proceed with filling in the rest of the form. Unless I've switched to my mouse and actually clicked on the flag, I still might not realize that the flag is somehow an integral part of the phone number field.

Yeah, that's a valid point.
Maybe the design of the phone number input isn't ideal by itself.
But I guess every application out there would have its own idea of the "best" design and there's no single "one size fits all" solution.
The default design of this component assumes the user is a young person having much experience online.
Other apps may have different potential audience.
For grandmas maybe a dedicated "Phone number country" select could be an option, or maybe not.
It can be implemented currently using a "Without country select" phone number input component and implementing a custom country select.




At this point I gave up re-reading this thread.

**Comment by catamphetamine** [comment]
@TylerRick Either the subject of our discussion is too complex or we're both too talkative.

**Comment by catamphetamine** [comment]
@TylerRick 
Created two subtasks from this issue: https://github.com/catamphetamine/react-phone-number-input/issues/287 and https://github.com/catamphetamine/react-phone-number-input/issues/288

A subtask not created: "Add `switchToNationalNumberModeAfterDetectingCountry` property".

When the country is not selected and the phone number input is therefore in "international format" input mode, a user starts typing an international phone number and as soon as the country is determined from the number that number could be converted to a "national" format.
For example, if a user enters `+1 234 567 8900` and US is selected then the input value becomes `(234) 567-8900`.
But wouldn't it confuse the user?
I guess it would: a person may not want that at all and may be comfortable enough with the international number format.
I guess I won't be implemeting this `switchToNationalNumberModeAfterDetectingCountry` for this reason: I don't find it useful for that reason.

**Comment by catamphetamine** [comment]
@TylerRick 
And another one: https://github.com/catamphetamine/react-phone-number-input/issues/289
That one is just a "notes to self for some future maybe".
